[Store] 매장 생성 기능 구현

iop0217 opened the pull request:
### Description  
- 사장님만 매장을 등록할 수 있도록 권한 검사를 적용했습니다.  
- 매장은 최대 3개까지만 생성 가능하도록 제한을 두었습니다.  
- 요청 값 유효성 검증 및 예외 처리 로직을 적용했고, Postman을 통해 정상 동작을 확인했습니다.  
- 매장 생성 로직에 대한 테스트 코드를 작성하고, 정상 작동을 확인했습니다.
- **피드백 반영으로 일부 로직 위치 변경 및 예외 처리 방식 등을 수정했습니다.**
- **DTO 내부로 변환 메서드 이동 및 불필요한 StoreConverter 클래스를 제거하였습니다.**

---

### Changes  
- 매장 생성 API (`POST /api/stores`) 구현  
- `StoreRequestDto`, `StoreResponseDto` 작성 및 주석 추가  
- `GlobalException`과 `ErrorCode`를 활용한 예외 처리
- `message.properties`에 매장 관련 메시지 추가

---

### Screenshots  

#### ✅ 201 Created - 매장 생성 성공
![201](https://github.com/user-attachments/assets/3fbc8484-f95a-4ace-8a39-c24f0a20678a)

#### ❌ 400 Bad Request - 필수 입력 누락
![400](https://github.com/user-attachments/assets/af90acd8-be1c-4255-8014-73f0080037d9)

#### ❌ 401 Unauthorized - 인증되지 않은 사용자
![401](https://github.com/user-attachments/assets/f84ce832-6e4b-4c96-932e-495982b50576)

#### ❌ 403 Forbidden - 사장님 권한 아님
![403](https://github.com/user-attachments/assets/d63bf39c-9702-49dd-9a4b-d3829c475c44)

iop0217 committed:
Feature[Store]: 매장 생성 기능 구현

- StoreRequestDto, StoreResponseDto 생성
- StoreService 및 StoreController 작성
- 매장 최대 3개 제한 로직 포함
- GlobalException 및 ErrorCode 오타 수정 완료

iop0217 committed:
test [store]: 매장 생성 서비스 테스트 코드 추가

- 매장 정상 생성 흐름 테스트
- 저장 로직 및 DTO 변환 결과 검증

iop0217 committed:
Fix[Store]: 예외 처리 및 주석 정리

- 유효성 검사 추가 (@Valid)
- 사장님 권한 체크 로직
- 상태코드 명세 반영
- DTO/엔티티 주석 정리

iop0217 committed:
Merge branch 'dev' into feature/suhyun-stores

iop0217 committed:
fix [store]: 요청 메시지 처리 및 dev merge 대응 완료

- StoreRequestDto 관련 messages.properties 추가
- dev 병합 후 Postman 테스트로 JWT 인증 매장 생성 기능 정상 작동 확인
- JWT 인증 기반의 매장 생성 기능 정상 작동 확인

iop0217 committed:
Merge branch 'dev' into feature/suhyun-stores

iop0217 committed:
fix[store]: 테스트 코드 기대값 수정 및 정상 동작 확인

- 매장 생성 테스트에서 예상값 오타 수정
- 테스트 정상 통과 확인

anyubinn commented on src/main/java/com/delivery/igo/igo_delivery/api/store/controller/StoreController.java:
```diff
@@ -1,4 +1,41 @@
+            @Auth AuthUser loginUser,
+            @Valid @RequestBody StoreRequestDto requestDto
+    ) {
+        Users user = userRepository.findById(loginUser.getId())
```
findById와 같은 비즈니스 로직은 서비스 쪽에서 진행하는 것이 단일 책임 원칙에도 맞고 좋을 것 같아요!

iop0217 replied:
옳으신 말씀입니다.

anyubinn commented on src/main/java/com/delivery/igo/igo_delivery/api/store/service/StoreServiceImpl.java:
```diff
@@ -0,0 +1,45 @@
+        }
+
+        // 사장이 이미 3개의 매장을 등록한 경우 예외 발생
+        long count = storeRepository.countByUsersAndStoreStatusIsNot(owner, com.delivery.igo.igo_delivery.api.store.entity.StoreStatus.CLOSED);
```
패키지 경로를 직접적으로 명시하는 것 보다 import해서 사용하는 것이 가독성 측면에서도 더 나을 것 같아요!

iop0217 replied:
전 바보에요

anyubinn commented on src/main/java/com/delivery/igo/igo_delivery/api/store/service/StoreServiceImpl.java:
```diff
@@ -0,0 +1,45 @@
+        // 사장이 이미 3개의 매장을 등록한 경우 예외 발생
+        long count = storeRepository.countByUsersAndStoreStatusIsNot(owner, com.delivery.igo.igo_delivery.api.store.entity.StoreStatus.CLOSED);
+        if (count >= 3) {
+            throw new IllegalStateException("매장은 최대 3개까지 운영할 수 있습니다.");
```
이 부분도 ErrorCode에 추가해서 GlobalException으로 받는 것이 일관적이고 좋을 것 같아요!

iop0217 replied:
반영했습니다

nagul2 commented on src/main/java/com/delivery/igo/igo_delivery/api/store/controller/StoreController.java:
```diff
@@ -1,4 +1,41 @@
+            @Auth AuthUser loginUser,
+            @Valid @RequestBody StoreRequestDto requestDto
+    ) {
+        Users user = userRepository.findById(loginUser.getId())
+                .orElseThrow(() -> new GlobalException(ErrorCode.USER_NOT_FOUND));
+
```
Controller에서 repository를 의존하는 것은 의존 계층이 깨지고, controller의 역할에 부합되지 않습니다.
해당 코드를 Service 계층으로 이동시긴 후 예외 전파를 통해 컨트롤러까지 예외가 발생되도록 해야 합니다.

컨트롤러의 역할은 뷰에 관련된것, 응답에 관련된것 이렇게 컨트롤러가 가진 역할에 대한 코드만 존재해야 한다고 생각해주시면 좋을 것 같습니다.

iop0217 replied:
제가 멍청해서 그래요...

nagul2 commented on src/main/java/com/delivery/igo/igo_delivery/api/store/converter/StoreConverter.java:
```diff
@@ -0,0 +1,44 @@
+import java.sql.Time;
+
+public class StoreConverter {
+
+    // 매장 생성 요청 DTO를 엔티티로 변환
+    public static Stores toEntity(StoreRequestDto dto, Users owner) {
+        return Stores.builder()
+                .storeName(dto.getStoreName())
+                .storeAddress(dto.getStoreAddress())
+                .storePhoneNumber(dto.getStorePhoneNumber())
+                .openTime(Time.valueOf(dto.getOpenTime()))
+                .endTime(Time.valueOf(dto.getEndTime()))
+                .minOrderPrice(dto.getMinOrderPrice())
+                .storeStatus(StoreStatus.LIVE)
+                .users(owner)
+                .reviewCount(0)
+                .avgRating(0.0)
+                .build();
+    }
+
+    // 매장 엔티티를 응답용 DTO로 변환
+    public static StoreResponseDto toDto(Stores store) {
+        return StoreResponseDto.builder()
+                .id(store.getId())
+                .storeName(store.getStoreName())
+                .storeAddress(store.getStoreAddress())
+                .storePhoneNumber(store.getStorePhoneNumber())
+                .openTime(store.getOpenTime().toLocalTime())
+                .endTime(store.getEndTime().toLocalTime())
+                .minOrderPrice(store.getMinOrderPrice())
+                .storeStatus(store.getStoreStatus().name())
+                .createdAt(store.getCreatedAt())
+                .modifiedAt(store.getModifiedAt())
+                .build();
+    }
+}
```
프로젝트 규모가 작어서 변환 메서드가 엄청 많이 생길 것 같지 않아 정적 팩토리 메서드는 각 Dto에 두는 것이 현재는 맞다고 생각되는데,
이렇게 적용하신 배경이 있는지 궁금합니다.

iop0217 replied:
ERD 설계하면서 TIME과 LOCALTIME으로 분리되었기 때문에 그걸 변환시키려고 만들었습니다.
규모나 메서드 수와 관계없이 변환 메서드는 따로 두는 게 편리하지 않나 생각했습니다.
바꿀까요?

nagul2 replied:
말씀하신 목적이라면 Conver클래스를 두는 것보다 각 Dto안에 두어도 동일하게 관리할 수 있다고 생각은 들긴합니다만..
이부분은 그대로 두시고 목적성을 좀 더 알아보시고 팀원분들께 공유해주시면 다 같이 공부가 될 것 같습니다.

nagul2 commented on src/main/java/com/delivery/igo/igo_delivery/api/store/entity/Stores.java:
```diff
@@ -29,13 +28,13 @@ public class Stores extends BaseEntity {
-    @Column(nullable = false)
+    @Column(nullable = false, length = 30)
     private String storeName;
 
-    @Column(nullable = false)
+    @Column(nullable = false, length = 100)
     private String storeAddress;
 
-    @Column(nullable = false)
+    @Column(nullable = false, length = 20)
     private String storePhoneNumber;
```
ERD에는 따로 컬럼에 제한을 두진 않긴햇는데, 혹시 제한을 두어야 하는 거라면 ERD 수정이 필요할 것 같습니다.
그리고 bean validation이나 ServiceImpl 검증 코드가 필요할 것 같습니다.

iop0217 replied:
수정했습니다

nagul2 commented on src/main/java/com/delivery/igo/igo_delivery/api/store/service/StoreServiceImpl.java:
```diff
@@ -0,0 +1,45 @@
+    public StoreResponseDto createStore(StoreRequestDto requestDto, Users owner) {
+
+        // 사장님 권한이 아닌 경우 예외 발생
+        if (!owner.getUserRole().equals(UserRole.OWNER)) {
+            throw new GlobalException(ErrorCode.NOT_OWNER);
+        }
+
```
이부분은 제 PR이 머지되면 수정되는 코드로 반영해주시면 좋을 것 같습니다.

iop0217 replied:
반영했습니다

nagul2 commented on src/main/java/com/delivery/igo/igo_delivery/api/store/service/StoreServiceImpl.java:
```diff
@@ -0,0 +1,45 @@
+            throw new GlobalException(ErrorCode.NOT_OWNER);
+        }
+
+        // 사장이 이미 3개의 매장을 등록한 경우 예외 발생
+        long count = storeRepository.countByUsersAndStoreStatusIsNot(owner, com.delivery.igo.igo_delivery.api.store.entity.StoreStatus.CLOSED);
+        if (count >= 3) {
+            throw new IllegalStateException("매장은 최대 3개까지 운영할 수 있습니다.");
+        }
+
```
저희 IlligalStateException은 예외 응답 핸들러가 없어서 GlobalException으로 발생시키고 에러코드 입력해주셔야 합니다.
만약 저 에러 발생되면 필터에러 걸려서 예외 코드가 잘못 출력될 수 있어요

iop0217 replied:
아하 확인했습니다

nagul2 commented on src/main/java/com/delivery/igo/igo_delivery/common/exception/ErrorCode.java:
```diff
@@ -46,6 +46,9 @@ public enum ErrorCode {
     ROLE_CONSUMER_FORBIDDEN(HttpStatus.UNAUTHORIZED, "주문 고객이 아닙니다."),
     ROLE_OWNER_FORBIDDEN(HttpStatus.UNAUTHORIZED, "매장 사장님이 아닙니다."),
 
+    // Store
+    NOT_OWNER(HttpStatus.FORBIDDEN, "사장님 권한이 아닙니다."),
+
```
이부분은 제가 추가해놓은 코드가 있어서 dev 머지 해보셔야 할 것 같습니다.

iop0217 replied:
병합했습니다

nagul2 commented on src/test/java/com/delivery/igo/igo_delivery/api/store/service/StoreServiceTest.java:
```diff
@@ -0,0 +1,78 @@
+                .nickname("초코라떼")
+                .userRole(UserRole.OWNER)
+                .build();
+
+        StoreRequestDto request = new StoreRequestDto(
+                "초코라떼가제일좋아",
+                "인천시 미추홀구",
+                "010-8282-8282",
+                LocalTime.of(8, 0),
+                LocalTime.of(22, 0),
+                10000
+        );
+
+        // 이미 등록된 매장이 2개라고 가정
+        when(storeRepository.countByUsersAndStoreStatusIsNot(owner, StoreStatus.CLOSED))
+                .thenReturn(2L);
+
+        // storeRepository.save() 호출 시, 실제 매장 엔티티 리턴하도록 설정
+        Stores store = Stores.builder()
+                .users(owner)
+                .storeName("초코라떼가제일좋아")
+                .storeAddress("인천시 미추홀구")
+                .storePhoneNumber("010-8282-8282")
+                .openTime(Time.valueOf(request.getOpenTime()))
+                .endTime(Time.valueOf(request.getEndTime()))
+                .minOrderPrice(request.getMinOrderPrice())
+                .storeStatus(StoreStatus.LIVE)
+                .build();
+
+        when(storeRepository.save(any(Stores.class))).thenReturn(store);
+
+        // when
+        var response = storeService.createStore(request, owner);
+
+        // then
+        assertThat(response).isNotNull();
+        assertThat(response.getStoreName()).isEqualTo("초코라떼가제일좋아");
+        verify(storeRepository).save(any(Stores.class));
+    }
+}
```
비정상 케이스도 몇개 테스트 해봐주시면 좋을 것 같습니다.

iop0217 replied:
다음 API부터 하겠습니다
지금 이걸 하기에는 시간이 부족할 것 같아요

iop0217 committed:
fix[store]: 리뷰 피드백 반영 (컨트롤러 책임 분리, 예외 통일, 유효성 검증 정리 등)

- Controller의 사용자 조회 로직을 Service로 이동
- IllegalStateException -> GlobalException으로 변경 및 ErrorCode 수정
- StoreStatus import 경로 정리
- Entity 내 length 속성 제거 및 DTO에 @Size 유효성 검증 추가

iop0217 committed:
Merge branch 'dev' into feature/suhyun-stores

anyubinn commented on src/main/java/com/delivery/igo/igo_delivery/api/store/controller/StoreController.java:
```diff
@@ -1,4 +1,35 @@
-}
+
+    private final StoreService storeService;
+    private final UserRepository userRepository;
```
사용하지 않는 UserRepository는 제거해 주시면 좋을 것 같아요!

iop0217 replied:
코드 정리를 까먹었나봐요! 수정했습니다

iop0217 committed:
refactoring[store]: 사용하지 않는 코드 정리

iop0217 committed:
refactoring[store]: StoreConverter 제거 및 DTO 내부 변환 메서드로 이동

- 매장 생성 요청 DTO에 toEntity(owner) 메서드 추가
- 매장 응답 DTO에 from(store) 정적 팩토리 메서드 추가
- StoreConverter 삭제 및 서비스 내 코드 수정

rlqor9688 reviewed (approved):
고생하셨습니다!
